src/client/Client.cc

```diff
@@ -3783,47 +3805,63 @@ void Client::flush_caps()
   }
 }
 
-void Client::flush_caps(Inode *in, MetaSession *session)
+void Client::flush_caps(Inode *in, MetaSession *session, bool kick)
 {
   ldout(cct, 10) << "flush_caps " << in << " mds." << session->mds_num << dendl;
   Cap *cap = in->auth_cap;
   assert(cap->session == session);
 
-  send_cap(in, session, cap, get_caps_used(in), in->caps_wanted(),
-	   (cap->issued | cap->implemented), in->flushing_caps);
+  int flushing = in->flushing_caps;
+  map<ceph_tid_t,int> flushes;
+  for (map<ceph_tid_t,int>::reverse_iterator p = in->flushing_cap_tids.rbegin();
+       p != in->flushing_cap_tids.rend();
+       ++p) {
+    if (kick && flush_tids_to_kick.count(p->first) == 0)
+      continue;
+    flushes[p->first] = p->second;
+    // no need to re-send all pending flushes
+    flushing &= ~p->second;
+    if (!flushing)
+      break;
```

Comment:
Doesn't this break eg handle_cap_flush_ack when it clears out all flushes older than the acked one? Or wait_sync_caps?